README.md
# @alpler/logger

## Table of Contents
- [Install](#install)
- [Introduction](#introduction)
- [Log Levels](#log-levels)

## Install
This is a NodeJs Module, it can be installed through the npm registry.

To install use the following command in your root directory.

    $ npm install @alpler/logger

## Introduction
This is a simple logger for NodeJs. It will generate to log Files in which messages can be logged. 
	errors.log	Will only include messages with a level higher than WARN 
	log.log         Will include all messages above the before set Log Level

Simple example of how use it

JavaScript:

    const logger = require("@alpler/logger");
    logger.setLogLevel(logger.LogLevel.INFO);
    logger.log(logger.LogLevel.ERROR, "example.js:5", "There is no code left to execute :(");


TypeScript: 

    import * as logger from "@alpler/logger"
    
    logger.setLogLevel(logger.LogLevel.INFO)
    logger.log(logger.LogLevel.ERROR, "example.ts:3", "There is no code left to execute :( ") 

	
	
OUTPUT:

    [21.02.2022] [01:08] ERROR example.ts:3 There is no code left to execute :(


## Log Levels
Each Log Level describes the different importance of a message. 

### Types:

In the version 1.2.4 there are six different Log levels these are


Name | Value | Description
---------|----------|---------
 FATAL | 5000 | Servere errors that might cause a crash
 ERROR | 4000 | Erros which may prevent normal execution of the program
 WARN | 3000 | Potential problems which could also just be nothing
 INFO | 2000 | Information about the progress of the program
 DEBUG | 1000 | Detailed tracing messages for the progress of the program
 ALL | 0 | All that could you ever want to log



### Usages:

#### **Setting the Log Level**

Typescript:

    logger.setLogLevel(logger.LogLevel.INFO)

JavaScript:

    logger.setLogLevel(logger.LogLevel.INFO)
   
   This will set the Level at which the logger will start writing it into the log.log and also the console

#### **Logging a message**

Typescript:

    logger.log(logger.LogLevel.ERROR, "example.ts:3", "There is no code left to execute :(")

JavaScript:

    logger.log(logger.LogLevel.ERROR, "example.js:5", "There is no code left to execute :(")

In this Case we pass a Log Level with a message so the logger can check if the message should be logged.



changelog.md
# Changelog

This file contains all changes to the project from version 1.2.1 till 1.2.1

## v1.9.1
* Added different Log Levels and deleted logError Function
* Added ability to set level at which logging should start

## v1.2.1 (20.03.2022)
* Switched to Namespaces, Reason: TypeScript didn't like the old import way

index.js
"use strict";
// Type definitions for @alpler/logger
// Project: https://github.com/Aelpler/-alpler-logger
// Author: https://github.com/Aelpler
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.log = exports.deleteFiles = exports.createFiles = exports.setLogLevel = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const logLevel_1 = require("./logLevel");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return logLevel_1.LogLevel; } });
/**
 * Level at which a message will be logged
 * default: LogLevel.INFO
 */
let _logLevel = logLevel_1.LogLevel.INFO;
/**
 * Folder in which logs will be saved
 */
let _logFolder = "logs";
/**
 * File where all messages will end up
 */
let _log_File = "log.log";
/**
 * File where messages with level greater than LogLevel.WARN will end up
 */
let _log_Error_File = "errors.log";
/**
 * Set log level at which it will log messages
 * @param {LogLevel} logLevel How important a message needs to be, to be loged
 * @public
 */
function setLogLevel(logLevel) {
    _logLevel = logLevel;
}
exports.setLogLevel = setLogLevel;
/**
 * Create local files in which you log messages will end up.
 * Folder in which log.log and erros.log will be
 * @public
 */
function createFiles() {
    if (!fs_1.default.existsSync(path_1.default.join("./", _logFolder)))
        fs_1.default.mkdirSync(path_1.default.join("./", _logFolder));
    if (!fs_1.default.existsSync(path_1.default.join("./", _logFolder, _log_File)))
        fs_1.default.writeFileSync(path_1.default.join("./", _logFolder, _log_File), "");
    if (!fs_1.default.existsSync(path_1.default.join("./", _logFolder, _log_Error_File)))
        fs_1.default.writeFileSync(path_1.default.join("./", _logFolder, _log_Error_File), "");
}
exports.createFiles = createFiles;
/**
 * Delete local files with log messages
 * In order to continue without messages don't forget to use createFiles() after
 * @public
 */
function deleteFiles() {
    if (fs_1.default.existsSync(path_1.default.join("./", _logFolder)))
        fs_1.default.rmSync(path_1.default.join("./", _logFolder), { recursive: true, force: true });
}
exports.deleteFiles = deleteFiles;
/**
 * Log messages into file and also into console
 * @param {LogLevel} logLevel Imporatance of you message
 * @param {string} location Where in the program a message came from
 * @param {string} text The message you want to log
 * @public
 */
function log(logLevel, location, text) {
    let formatedLog = formatText(logLevel, location, text);
    // If logLevel is greater than error write also to error.log
    if (logLevel >= 3000) {
        fs_1.default.appendFile(path_1.default.join("./", _logFolder, _log_Error_File), formatedLog, function (err) {
            if (err)
                console.log(`Couldn't write to error log ${err}`);
        });
    }
    // Logs if logLevel from message is greater or simular to set _logLevel
    if (_logLevel <= logLevel) {
        fs_1.default.appendFile(path_1.default.join("./", _logFolder, _log_File), formatedLog, function (err) {
            if (err)
                console.log(`Couldn't write to log file ${err}`);
        });
    }
    if (logLevel >= 3000 || _logLevel <= logLevel) {
        console.log(formatedLog);
    }
}
exports.log = log;
/**
 * Formats the given parameters into single text so all have the same format
 * @param {LogLevel} logLevel Imporatance of you message
 * @param {string} location Where in the program a message came from
 * @param {string} text The message you want to log
 * @return {string} Formated log message with date,time,level,location,text
 */
function formatText(logLevel, location, text) {
    let date = new Date();
    let dateString = `[${formatNumber(date.getDate())}.${formatNumber(date.getMonth())}.${date.getFullYear()}]`;
    let timeString = `[${formatNumber(date.getHours())}:${formatNumber(date.getMinutes())}]`;
    return `${dateString} ${timeString} ${logLevel_1.LogLevel[logLevel]} ${location} ${text}\n`;
}
/**
 * Converts a number to string, adds n is below value of 10
 * @param {number} n Number which will be converted to string
 * @return {string} Formated number, with may added 0
 */
function formatNumber(n) {
    return (n < 10 ? '0' : '') + n;
}


src/index.ts
// Type definitions for @alpler/logger
// Project: https://github.com/Aelpler/-alpler-logger
// Author: https://github.com/Aelpler

import fs from "fs"
import path from "path"
import { LogLevel } from "./logLevel"

namespace logger {

    /**
     * Level at which a message will be logged
     * default: LogLevel.INFO
     */
    let _logLevel: LogLevel = LogLevel.INFO

    /**
     * Folder in which logs will be saved
     */
    let _logFolder = "logs"
    /**
     * File where all messages will end up
     */
    let _log_File = "log.log"
    /**
     * File where messages with level greater than LogLevel.WARN will end up
     */
    let _log_Error_File = "errors.log"

    /**
     * Set log level at which it will log messages
     * @param {LogLevel} logLevel How important a message needs to be, to be loged
     * @public
     */
    export function setLogLevel(logLevel: LogLevel): void {
        _logLevel = logLevel
    }

    /**
     * Create local files in which you log messages will end up.
     * Folder in which log.log and erros.log will be
     * @public
     */
    export function createFiles(): void {
        if (!fs.existsSync(path.join("./", _logFolder)))
            fs.mkdirSync(path.join("./", _logFolder))

        if (!fs.existsSync(path.join("./", _logFolder, _log_File)))
            fs.writeFileSync(path.join("./", _logFolder, _log_File), "")
        if (!fs.existsSync(path.join("./", _logFolder, _log_Error_File)))
            fs.writeFileSync(path.join("./", _logFolder, _log_Error_File), "")
    }

    /**
     * Delete local files with log messages
     * In order to continue without messages don't forget to use createFiles() after
     * @public 
     */
    export function deleteFiles(): void {
        if (fs.existsSync(path.join("./", _logFolder)))
            fs.rmSync(path.join("./", _logFolder), { recursive: true, force: true })
    }

    /**
     * Log messages into file and also into console
     * @param {LogLevel} logLevel Imporatance of you message
     * @param {string} location Where in the program a message came from 
     * @param {string} text The message you want to log
     * @public
     */
    export function log(logLevel: LogLevel, location: string, text: string): void {
        let formatedLog: string = formatText(logLevel, location, text)

        // If logLevel is greater than error write also to error.log
        if (logLevel >= LogLevel.WARN) {
            fs.appendFile(path.join("./", _logFolder, _log_Error_File), formatedLog, function (err: any) {
                if (err)
                    console.log(`Couldn't write to error log ${err}`)
            })
        }

        // Logs if logLevel from message is greater or simular to set _logLevel
        if (_logLevel <= logLevel) {
            fs.appendFile(path.join("./", _logFolder, _log_File), formatedLog, function (err: any) {
                if (err)
                    console.log(`Couldn't write to log file ${err}`)
            })
        }

        if (logLevel >= 3000 || _logLevel <= logLevel) {
            console.log(formatedLog)
        }
    }

    /**
     * Formats the given parameters into single text so all have the same format
     * @param {LogLevel} logLevel Imporatance of you message
     * @param {string} location Where in the program a message came from 
     * @param {string} text The message you want to log
     * @return {string} Formated log message with date,time,level,location,text
     */
    function formatText(logLevel: LogLevel, location: string, text: string): string {
        let date = new Date()
        let dateString = `[${formatNumber(date.getDate())}.${formatNumber(date.getMonth())}.${date.getFullYear()}]`
        let timeString = `[${formatNumber(date.getHours())}:${formatNumber(date.getMinutes())}]`
        return `${dateString} ${timeString} ${logLevel.toLocaleString()} ${location} ${text}\n`
    }

    /**
     * Converts a number to string, adds n is below value of 10
     * @param {number} n Number which will be converted to string
     * @return {string} Formated number, with may added 0 
     */
    function formatNumber(n: number): string {
        return (n < 10 ? '0' : '') + n;
    }

}

export { logger, LogLevel }


src/@types/index.d.ts
// Type definitions for @alpler/logger
// Project: https://github.com/Aelpler/-alpler-logger
// Author: https://github.com/Aelpler

import { LogLevel } from "../logLevel";

declare module "@alpler/logger" {

    export namespace logger {
        /**
         * Log messages into file and also into console
         * @param {LogLevel} logLevel Imporatance of you message
         * @param {string} location Where in the program a message came from 
         * @param {string} text The message you want to log
         * @public
         */
        export function log(logLevel: LogLevel, location: string, text: string): void

        /**
         * Set log level at which it will log messages
         * @param {LogLevel} logLevel How important a message needs to be, to be loged
         * @public
         */
        export function setLogLevel(logLevel: LogLevel): void

        /**
         * Create local files in which you log messages will end up.
         * Folder in which log.log and erros.log will be
         * @public
         */
        export function createFiles(): void

        /**
         * Delete local files with log messages
         * In order to continue without messages don't forget to use createFiles() after
         * @public 
         */
        export function deleteFiles(): void

        /**
         * Importance of a message
         * FATAL: Servere errors that might cause a crash
         * ERROR: Erros which may cause that the program will not be executed as intended
         * WARN:  Potential problems which could also just be nothing
         * INFO:  Information about the progress of the program
         * DEBUG: Detailed tracing messages for the progress of program
         * ALL:   All messages that did go through logging function
         * @return {number} Value of Log Level as number
         */

    }
}
export { LogLevel } from "../logLevel"

logLevel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    /** Servere errors that might cause a crash */
    LogLevel[LogLevel["FATAL"] = 5000] = "FATAL";
    /** Erros which may prevent normal execution of the program */
    LogLevel[LogLevel["ERROR"] = 4000] = "ERROR";
    /** Potential problems which could also just be nothing */
    LogLevel[LogLevel["WARN"] = 3000] = "WARN";
    /** Information about the progress of the program */
    LogLevel[LogLevel["INFO"] = 2000] = "INFO";
    /** Detailed tracing messages for the progress of the program */
    LogLevel[LogLevel["DEBUG"] = 1000] = "DEBUG";
    /** All messages that did go through logging function */
    LogLevel[LogLevel["ALL"] = 0] = "ALL";
})(LogLevel || (LogLevel = {}));
exports.LogLevel = LogLevel;
